File: TappyColour/Model/PaintingSession.swift
import Foundation
import SwiftUI
import Combine

@MainActor
final class PaintingSession: ObservableObject {

    @Published private(set) var grid: PixelGrid
    @Published private(set) var palette: Palette
    @Published var selectedPaletteIndex: Int
    @Published private(set) var isComplete: Bool = false
    @Published var numbersVisible: Bool = true

    // Single-step undo: stores the (col, row) pairs changed by the last action
    private var undoStack: [[(col: Int, row: Int)]] = []   // max 1 entry

    // Tracks cells painted in the current drag gesture to avoid repainting
    private var currentDragPainted: Set<CellCoord> = []
    private var isDragging: Bool = false

    // Cache: total cells per palette index (constant after conversion)
    private let totalCellsPerColour: [Int]

    let artworkID: UUID

    // Injected save callback — called after every mutating action
    var onNeedsSave: ((PaintingSession) -> Void)?

    init(artwork: Artwork) {
        self.artworkID = artwork.id
        self.grid = artwork.grid
        self.palette = artwork.palette
        // Start on the first incomplete colour
        self.selectedPaletteIndex = (0..<artwork.palette.colours.count)
            .first(where: { artwork.grid.unpaintedPerColour[$0] > 0 }) ?? 0

        // Compute total cells per colour (constant)
        var totals = [Int](repeating: 0, count: artwork.palette.colours.count)
        for cell in artwork.grid.cells {
            let idx = Int(cell.paletteIndex)
            if idx < totals.count { totals[idx] += 1 }
        }
        self.totalCellsPerColour = totals
    }

    // MARK: - Query helpers for PaletteStripView

    func totalCells(forIndex index: Int) -> Int {
        guard index < totalCellsPerColour.count else { return 0 }
        return totalCellsPerColour[index]
    }

    func paintedCells(forIndex index: Int) -> Int {
        guard index < totalCellsPerColour.count,
              index < grid.unpaintedPerColour.count else { return 0 }
        return totalCellsPerColour[index] - grid.unpaintedPerColour[index]
    }

    // MARK: - Painting tools

    func tap(col: Int, row: Int) {
        // Only paint if the cell's colour matches the selected colour
        let cell = grid.cell(col: col, row: row)
        guard Int(cell.paletteIndex) == selectedPaletteIndex, !cell.painted else { return }

        if grid.paint(col: col, row: row) {
            undoStack = [[(col: col, row: row)]]
            afterPaint()
        }
    }

    func dragBegan() {
        isDragging = true
        currentDragPainted.removeAll()
    }

    func dragMoved(col: Int, row: Int) {
        guard isDragging else { return }
        let coord = CellCoord(col: col, row: row)
        guard !currentDragPainted.contains(coord) else { return }

        let cell = grid.cell(col: col, row: row)
        guard Int(cell.paletteIndex) == selectedPaletteIndex, !cell.painted else { return }

        if grid.paint(col: col, row: row) {
            currentDragPainted.insert(coord)
        }
    }

    func dragEnded() {
        isDragging = false
        if !currentDragPainted.isEmpty {
            let changed = currentDragPainted.map { (col: $0.col, row: $0.row) }
            undoStack = [changed]
            currentDragPainted.removeAll()
            afterPaint()
        }
    }

    func bucketFill(col: Int, row: Int) {
        let targetCell = grid.cell(col: col, row: row)
        guard Int(targetCell.paletteIndex) == selectedPaletteIndex,
              !targetCell.painted else { return }

        let region = connectedRegion(startCol: col, startRow: row, paletteIndex: targetCell.paletteIndex)
        let changed = grid.paintCells(region)
        if !changed.isEmpty {
            undoStack = [changed]
            afterPaint()
        }
    }

    func fillAllOfNumber() {
        let indices = grid.unpaintedIndices(forPaletteIndex: UInt8(selectedPaletteIndex))
        let coords = indices.map { idx -> (col: Int, row: Int) in
            (col: idx % grid.columns, row: idx / grid.columns)
        }
        let changed = grid.paintCells(coords)
        if !changed.isEmpty {
            undoStack = [changed]
            afterPaint()
        }
    }

    // MARK: - Undo

    func undo() {
        guard !isDragging, let last = undoStack.last else { return }
        grid.unpaintCells(last)
        undoStack.removeLast()
        isComplete = false
        numbersVisible = true
        onNeedsSave?(self)
    }

    // MARK: - Palette editing

    func replacePaletteColour(at index: Int, with rgba: RGBA8) {
        palette.replaceColour(at: index, with: rgba)
        onNeedsSave?(self)
    }

    func resetPalette() {
        palette.reset()
        onNeedsSave?(self)
    }

    // MARK: - Private helpers

    private func afterPaint() {
        autoAdvanceIfNeeded()
        checkAndHandleCompletion()
        onNeedsSave?(self)
    }

    private func autoAdvanceIfNeeded() {
        guard selectedPaletteIndex < grid.unpaintedPerColour.count,
              grid.unpaintedPerColour[selectedPaletteIndex] == 0 else { return }

        let count = palette.colours.count
        let next = (1..<count)
            .map { (selectedPaletteIndex + $0) % count }
            .first(where: { $0 < grid.unpaintedPerColour.count && grid.unpaintedPerColour[$0] > 0 })
        if let next { selectedPaletteIndex = next }
    }

    private func checkAndHandleCompletion() {
        guard grid.isComplete, !isComplete else { return }
        isComplete = true
        Task { @MainActor in
            try? await Task.sleep(nanoseconds: 500_000_000)
            withAnimation(.easeOut(duration: 0.8)) {
                numbersVisible = false
            }
        }
    }

    private func connectedRegion(startCol: Int, startRow: Int, paletteIndex: UInt8) -> [(col: Int, row: Int)] {
        var queue: [(Int, Int)] = [(startCol, startRow)]
        var visited = Set<CellCoord>()
        var result: [(col: Int, row: Int)] = []

        while !queue.isEmpty {
            let (c, r) = queue.removeFirst()
            guard c >= 0, c < grid.columns, r >= 0, r < grid.rows else { continue }
            let coord = CellCoord(col: c, row: r)
            guard !visited.contains(coord) else { continue }
            let cell = grid.cell(col: c, row: r)
            guard cell.paletteIndex == paletteIndex, !cell.painted else { continue }

            visited.insert(coord)
            result.append((col: c, row: r))
            queue.append(contentsOf: [(c+1, r), (c-1, r), (c, r+1), (c, r-1)])
        }
        return result
    }
}

// MARK: - Supporting types

private struct CellCoord: Hashable {
    let col: Int
    let row: Int
}

File: web/src/model/PaintingSession.ts
import { EventEmitter } from '../util/events'
import type { PixelGrid } from '../engine/PixelGrid'
import type { Palette } from '../engine/Palette'
import type { RGBA8, CellCoord } from '../engine/types'
import { cellKey } from '../engine/types'
import { floodFillRegion, cellsForNumber } from '../engine/floodFill'
import type { ArtworkStore } from './ArtworkStore'
import type { Artwork } from './Artwork'
import type { PaintTool } from './PaintTool'
export type { PaintTool } from './PaintTool'

export type PaintEventType =
  | 'gridChanged'
  | 'paletteChanged'
  | 'selectionChanged'
  | 'completionChanged'
  | 'numbersVisibilityChanged'
  | 'toolChanged'
  | 'colourCompleted'
  | 'hintRequested'
  | 'wrongCell'

export class PaintingSession extends EventEmitter<{ change: PaintEventType }> {
  readonly artworkID: string
  grid: PixelGrid
  palette: Palette
  selectedPaletteIndex: number
  isComplete = false
  numbersVisible = true
  currentTool: PaintTool = 'tap'
  lastCompletedPaletteIndex: number | null = null
  lastRejectedCell: CellCoord | null = null

  private undoStack: CellCoord[][] = [] // max 1 entry
  private currentDragPainted = new Set<number>() // cellKey values
  private isDragging = false
  private totalCellsPerColour: number[]
  private numbersVisibilityTimeout: ReturnType<typeof setTimeout> | null =
    null
  private originalArtwork: Artwork

  constructor(artwork: Artwork, private store: ArtworkStore) {
    super()

    this.originalArtwork = artwork
    this.artworkID = artwork.id
    this.grid = artwork.grid
    this.palette = artwork.palette
    this.isComplete = artwork.isComplete

    // Compute totalCellsPerColour based on palette size
    this.totalCellsPerColour = new Array(artwork.palette.colours.length).fill(0)

    // Count total cells for each palette index
    for (let col = 0; col < artwork.grid.columns; col++) {
      for (let row = 0; row < artwork.grid.rows; row++) {
        const cell = artwork.grid.cell(col, row)
        if (cell.paletteIndex < this.totalCellsPerColour.length) {
          this.totalCellsPerColour[cell.paletteIndex]++
        }
      }
    }

    this.currentTool = artwork.sessionState?.currentTool ?? 'tap'

    // Find first incomplete colour unless valid persisted state says where the child left off.
    const restoredPaletteIndex = artwork.sessionState?.selectedPaletteIndex
    if (this.validPaletteIndex(restoredPaletteIndex)) {
      this.selectedPaletteIndex = restoredPaletteIndex
      return
    }

    this.selectedPaletteIndex = 0

    for (let i = 0; i < artwork.palette.colours.length; i++) {
      if (artwork.grid.unpaintedForColour(i) > 0) {
        this.selectedPaletteIndex = i
        break
      }
    }
  }

  private validPaletteIndex(index: number | undefined): index is number {
    return index !== undefined && index >= 0 && index < this.palette.colours.length
  }

  /**
   * Get total number of cells for a palette colour.
   */
  totalCells(index: number): number {
    return this.totalCellsPerColour[index] || 0
  }

  /**
   * Get number of painted cells for a palette colour.
   */
  paintedCells(index: number): number {
    const total = this.totalCells(index)
    const unpainted = this.grid.unpaintedForColour(index)
    return total - unpainted
  }

  /**
   * Paint a single cell at (col, row) if it matches selectedPaletteIndex and is unpainted.
   */
  tap(col: number, row: number): void {
    const cell = this.grid.cell(col, row)

    if (cell.paletteIndex !== this.selectedPaletteIndex || cell.painted) {
      if (!cell.painted) this.rejectCell(col, row)
      return
    }

    this.grid.paint(col, row)
    this.undoStack = [[{ col, row }]]
    this.emit('change', 'gridChanged')
    this.afterPaint()
  }

  /**
   * Switch the active painting tool. Aborts any in-progress drag.
   */
  setTool(tool: PaintTool): void {
    if (this.currentTool === tool) return
    this.isDragging = false
    this.currentDragPainted.clear()
    this.currentTool = tool
    this.emit('change', 'toolChanged')
    this.saveToStore()
  }

  requestHint(): void {
    if (this.grid.unpaintedForColour(this.selectedPaletteIndex) === 0) return
    this.emit('change', 'hintRequested')
  }

  private rejectCell(col: number, row: number): void {
    this.lastRejectedCell = { col, row }
    this.emit('change', 'wrongCell')
  }

  /**
   * Switch the selected palette colour and persist the resumed state.
   */
  selectPaletteIndex(index: number): void {
    if (!this.validPaletteIndex(index) || this.selectedPaletteIndex === index) return
    this.selectedPaletteIndex = index
    this.emit('change', 'selectionChanged')
    this.saveToStore()
  }

  /**
   * Flood-fill the 4-connected region at (col, row) if it matches selectedPaletteIndex.
   */
  bucketFill(col: number, row: number): void {
    const cell = this.grid.cell(col, row)
    if (cell.paletteIndex !== this.selectedPaletteIndex || cell.painted) {
      if (!cell.painted) this.rejectCell(col, row)
      return
    }

    const coords = floodFillRegion(this.grid, col, row)
    if (coords.length === 0) return

    const painted = this.grid.paintCells(coords)
    if (painted.length === 0) return

    this.undoStack = [painted]
    this.emit('change', 'gridChanged')
    this.afterPaint()
  }

  /**
   * Fill every remaining cell of the selected palette colour.
   */
  fillAllOfSelected(): void {
    const coords = cellsForNumber(this.grid, this.selectedPaletteIndex)
    if (coords.length === 0) return

    const painted = this.grid.paintCells(coords)
    if (painted.length === 0) return

    this.undoStack = [painted]
    this.emit('change', 'gridChanged')
    this.afterPaint()
  }

  /**
   * Begin a drag operation. Clears the set of painted cells during this drag.
   */
  dragBegan(): void {
    this.isDragging = true
    this.currentDragPainted.clear()
  }

  /**
   * Move during a drag. Paint the cell if it matches selectedPaletteIndex, is unpainted,
   * and hasn't been painted in this drag.
   */
  dragMoved(col: number, row: number): void {
    if (!this.isDragging) return

    const key = cellKey(col, row)

    if (this.currentDragPainted.has(key)) {
      return
    }

    const cell = this.grid.cell(col, row)

    if (cell.paletteIndex !== this.selectedPaletteIndex || cell.painted) {
      return
    }

    this.grid.paint(col, row)
    this.currentDragPainted.add(key)
    this.emit('change', 'gridChanged')
  }

  /**
   * End a drag operation. Convert the painted cells set to coordinates and set undo stack.
   */
  dragEnded(): void {
    this.isDragging = false

    if (this.currentDragPainted.size === 0) {
      return
    }

    // Convert cellKey values back to coordinates
    const coords: CellCoord[] = []
    for (const key of this.currentDragPainted) {
      coords.push({
        col: key % 10000,
        row: Math.floor(key / 10000),
      })
    }

    this.undoStack = [coords]
    this.currentDragPainted.clear()

    this.afterPaint()
  }

  /**
   * Undo the last action.
   */
  undo(): void {
    if (this.isDragging || this.undoStack.length === 0) {
      return
    }

    this.grid.unpaintCells(this.undoStack[0])
    this.undoStack = []
    this.restoreIncompleteState()

    this.emit('change', 'gridChanged')
    this.saveToStore()
  }

  /**
   * Replace a palette colour.
   */
  replacePaletteColour(index: number, rgba: RGBA8): void {
    this.palette.replaceColour(index, rgba)
    this.emit('change', 'paletteChanged')
    this.saveToStore()
  }

  /**
   * Reset palette to original colours.
   */
  resetPalette(): void {
    this.palette.reset()
    this.emit('change', 'paletteChanged')
    this.saveToStore()
  }

  /**
   * Auto-advance to the next incomplete colour if current is complete.
   */
  private autoAdvanceIfNeeded(): void {
    if (this.grid.unpaintedForColour(this.selectedPaletteIndex) > 0) {
      return
    }

    // Find next colour with unpainted cells
    for (let i = this.selectedPaletteIndex + 1; i < this.palette.colours.length; i++) {
      if (this.grid.unpaintedForColour(i) > 0) {
        this.selectedPaletteIndex = i
        this.emit('change', 'selectionChanged')
        return
      }
    }
  }

  /**
   * Check if the grid is complete and handle completion state.
   */
  private checkAndHandleCompletion(): void {
    if (!this.grid.isComplete() || this.isComplete) {
      return
    }

    this.isComplete = true
    this.emit('change', 'completionChanged')

    this.clearNumbersVisibilityTimeout()

    // Delay 500ms, then start fade animation by hiding numbers
    this.numbersVisibilityTimeout = setTimeout(() => {
      this.numbersVisible = false
      this.emit('change', 'numbersVisibilityChanged')
      this.numbersVisibilityTimeout = null
    }, 500)
  }

  /**
   * Called after painting cells. Auto-advances colour, checks completion, and saves.
   */
  private afterPaint(): void {
    this.emitColourCompletedIfNeeded()
    this.autoAdvanceIfNeeded()
    this.checkAndHandleCompletion()
    this.saveToStore()
  }

  private emitColourCompletedIfNeeded(): void {
    if (this.grid.unpaintedForColour(this.selectedPaletteIndex) > 0) return
    if (this.totalCells(this.selectedPaletteIndex) === 0) return

    this.lastCompletedPaletteIndex = this.selectedPaletteIndex
    this.emit('change', 'colourCompleted')
  }

  destroy(): void {
    this.clearNumbersVisibilityTimeout()
  }

  toArtworkSnapshot(now = Date.now()): Artwork {
    return {
      id: this.artworkID,
      title: this.originalArtwork.title,
      source: this.originalArtwork.source,
      thumbnailBlob: this.originalArtwork.thumbnailBlob,
      sourceImageBlob: this.originalArtwork.sourceImageBlob,
      createdAt: this.originalArtwork.createdAt,
      lastModifiedAt: now,
      isComplete: this.isComplete,
      conversionSettings: this.originalArtwork.conversionSettings,
      palette: this.palette,
      grid: this.grid,
      sessionState: {
        currentTool: this.currentTool,
        selectedPaletteIndex: this.selectedPaletteIndex,
      },
    }
  }

  private restoreIncompleteState(): void {
    const wasComplete = this.isComplete
    const numbersWereHidden = !this.numbersVisible

    this.clearNumbersVisibilityTimeout()
    this.isComplete = false
    this.numbersVisible = true

    if (wasComplete) this.emit('change', 'completionChanged')
    if (numbersWereHidden) this.emit('change', 'numbersVisibilityChanged')
  }

  private clearNumbersVisibilityTimeout(): void {
    if (this.numbersVisibilityTimeout !== null) {
      clearTimeout(this.numbersVisibilityTimeout)
      this.numbersVisibilityTimeout = null
    }
  }

  /**
   * Build and save the current state to the artwork store.
   */
  private saveToStore(): void {
    const artwork = this.toArtworkSnapshot()

    this.store.save(artwork).catch(() => {
      // Silently fail on save error
    })
  }
}
